c_src/membrane_ffmpeg_swscale_plugin/scaler.c

```diff
@@ -0,0 +1,202 @@
+  if ((float)(output_height * input_width / input_height) <= output_width) {
+    // paddings on left and right
+
+    *scaled_width = output_height * input_width / input_height;
+    *scaled_height = output_height;
+
+    // both paddings need even width
+    *scaled_width = *scaled_width / 2 * 2;
+    if ((output_width - *scaled_width) % 4 != 0)
+      *scaled_width -= 2;
+
+    // only even size
+    *scaled_height = *scaled_height / 2 * 2;
+  } else {
+    // paddings above and below
+
+    *scaled_width = output_width;
+    *scaled_height = output_width * input_height / input_width;
+
+    // only even size
+    *scaled_width = *scaled_width / 2 * 2;
+
+    // both paddings need even height
+    *scaled_height = *scaled_height / 2 * 2;
+    if ((output_height - *scaled_height) % 4 != 0)
+      *scaled_height -= 2;
+  }
+}
+
+UNIFEX_TERM create(UnifexEnv *env, int input_width, int input_height,
+                   int output_width, int output_height) {
+  UNIFEX_TERM result;
+  State *state = unifex_alloc_state(env);
+  state->sws_context = NULL;
+
+  enum AVPixelFormat pixel_format = AV_PIX_FMT_YUV420P;
+  int scaling_algorithm = SWS_BILINEAR;
+
+  int scaled_width;
+  int scaled_height;
+
+  calculate_scaling_resolution(input_width, input_height, output_width,
+                               output_height, &scaled_width, &scaled_height);
+
+  state->input_width = input_width;
+  state->input_height = input_height;
+
+  state->output_width = output_width;
+  state->output_height = output_height;
+
+  state->scaled_width = scaled_width;
+  state->scaled_height = scaled_height;
+
+  state->sws_context = sws_getContext(input_width, input_height, pixel_format,
+                                      scaled_width, scaled_height, pixel_format,
+                                      scaling_algorithm, NULL, NULL, NULL);
+
+  if (!state->sws_context) {
+    result = create_result_error(env, "get_context");
+    goto exit_create;
+  }
+
+  if (av_image_alloc(state->scaled_data, state->scaled_linesize, scaled_width,
+                     scaled_height, pixel_format, 1) < 0) {
+    result = create_result_error(env, "scaled_alloc");
+    goto exit_create;
+  }
+
+  if (av_image_alloc(state->output_data, state->output_linesize, output_width,
+                     output_height, pixel_format, 1) < 0) {
+    result = create_result_error(env, "output_alloc");
+    goto exit_create;
+  }
+
+  result = create_result_ok(env, state);
+
+exit_create:
+  unifex_release_state(env, state);
+  return result;
+}
+
+void add_paddings(uint8_t *scaled_data[4], int scaled_width, int scaled_height,
+                  uint8_t *output_data[4], int output_width,
+                  int output_height) {
+  if (scaled_height == output_height) {
+    // paddings on left and right
+    int start = (output_width - scaled_width) / 2;
+    int end = start + scaled_width;
+
+    for (int h = 0; h < output_height; h++) {
```

Comment:
Add a comment about this for - it's for Y plan, right? Same goes for the next one